kedro-org/kedro-plugins
# Expand telemetry to better understand size of kedro projects in the wild

## Description

See #44, quotes related to the actual tasks related to the scope of this PR.

> To work out the size of a Kedro project, probably tracked per username:

- Number of datasets
- Number of nodes
- Number of pipelines

To work out how many team members are on a project, to show that Kedro helps with collaboration:

- Number of projects, identified by project_name
- A count of username per project


> **If possible add:** Size of projects by internal vs external users
> 
> **Nice to have:** Could we figure out what parts of a pipeline are data engineering vs data science specific?

> This issue was discussed as part of a **Technical Design** session:
> 
> * Size should be determined by the number of nodes in the default pipeline and the number of dataset entries in the catalog
> * Currently we use the `before_command_run` hook, but to send the size info to heap we'd also need to either use the `before_pipeline_run` or `after_context_created` hook.
> * To be decided by the person implementing this: inside `kedro-telemetry` would we do a call to the Heap API for every Kedro hook that's called, or do we aggregate the data fetched in the hooks and then send it to Heap in one call?
> 
> 1. Need to implement the size calculation and sending it to heap
> 2. Make a "dashboard" or other view on Heap that shows the Kedro project size


## Development notes


* Number of pipelines
* Number of nodes in the `__default__` pipeline
* Number of datasets


### Setup for HEAP testing environment
* https://confluence.quantumblack.com/display/CAI/Heap+Analytics+for+Kedro-Viz?searchId=ZC01QX5OV

# demo
![image](https://user-images.githubusercontent.com/18221871/193869216-2f7a410b-1065-4eb4-a3fd-2c856892dae1.png)

# How to review?
There are mainly 2 ways that users run a project
* `kedro run`
* Manually create a `KedroSessions` and do `session.run`

The difference is - the first one will always invoke the CLI first, and currently this is the only hook that captures the project metadata. `PROJECT_PROPERTIES` is a global variable that tracks this information. So the order is always CLI hook -> Kedro's run hook.

However, it's also common to skip the CLI (Databricks or any interactive workflow), and I think it's important to capture the statistics for these projects. Without the project metadata, the data on HEAP will be inconsistent, so a manual `bootstrap_project` call is added to retrieve the `metadata`. 

# How to test?
* `kedro new -s pandas-iris` # telemetry
* `cd telemetry`

1. `kedro run`
2. `kedro ipython` - `session.run()`

You should see events on Heap immediately (Better to use Dev environment) - check the `Live Data Feed` page on Heap. If you have trouble with the setup, feel free to DM me I can show you quickly how it's done.

## Checklist

- [ ] Added a description of this change in the relevant `RELEASE.md` file
